### DISCUSSION TITLE
Files management in DB

### DISCUSSION THREAD

**Discussion by rbovard** [QGEP/QGEP#780] [Q&A] [answered]
Hello,

I have some remarks about the files management.

1. Is it intentional that there is no foreign key for `file.fk_data_media`?
2. There is a [function](https://github.com/QGEP/datamodel/blob/master/view/vw_file.sql#L62) that convert `\` to `/` in the path on `file` insert, but there is no same logical on `data_media` insert. So if we use `\` in `data_media.path`, we cannot insert data through `vw_file` because there is no match [here](https://github.com/QGEP/datamodel/blob/master/view/vw_file.sql#L93) between the path defined in `data_media` and the one set in `vw_file`.
3. There is an issue in the `vw_file`. Since data_media.path can be NULL (see [field comment](https://github.com/QGEP/datamodel/blob/master/09_qgep_dictionaries_kek.sql#L58)), `vw_file._url` will be null in this case.

Thanks in advance for you input about these points.

**Comment by sjib** [comment]
The foreign key is defined in the datamodel:
https://github.com/QGEP/datamodel/blob/e94fef6296354464e5855be98275596af0b0c3aa/04_vsa_kek_extension.sql#L193

But the relation that should go with the fk is missing:
ALTER TABLE qgep_od.file ADD CONSTRAINT rel_file_data_media FOREIGN KEY (fk_data_media) REFERENCES qgep_od.data_media(obj_id) ON UPDATE CASCADE ON DELETE set null;

**Reply by sjib** [reply]
Here is a patch: https://github.com/QGEP/datamodel/pull/193

@rbovard Can you check if this solves part of your problem?

**Reply by rbovard** [reply]
Thanks, I've answered in the PR.

**Comment by sjib** [comment, marked as answer]
> There is a [function](https://github.com/QGEP/datamodel/blob/master/view/vw_file.sql#L62) that convert \ to / in the path on file insert, but there is no same logical on data_media insert. So if we use \ in data_media.path, we cannot insert data through vw_file because t[here](https://github.com/QGEP/datamodel/blob/master/view/vw_file.sql#L93) is no match here between the path defined in data_media and the one set in vw_file.

As far as I understand the concept of this view `vw_file` it is, that you do not have to fill in fk_data_media yourself, but that you can just copy/paste the whole filepath of your file and it looks up, whether there is a data_media, that fits the beginning of that path and then cuts of the path of the data_media and just inserts the relative_path in file.

Yes and it is correct - data_media.path has to be also with `/` , else it will not find that data_media when you try to insert e.g.

-- file.class = Manhole (3858)
-- file.identifier = '8486_0001.jpg'
-- file.kind = photo (3772)
-- file.object = 'test'
-- file._url = 'C:\Daten\QGEP\casmedia\foto'
`INSERT INTO qgep_od.vw_file (identifier, kind, object, _url) VALUES ('8486_0001.jpg', 3772, 'test', 'C:\Daten\QGEP\casmedia\foto')`

![20221208_table_data_media](https://user-images.githubusercontent.com/1265865/206460209-076e8a26-ce34-4e83-96b6-36f6ed4ff899.png)
![20221208_table_file_after_insert](https://user-images.githubusercontent.com/1265865/206460231-27858167-0b05-4937-930d-f1c66e101d70.png)

If data_media.pat is with `\`
![20221208_table_data_media2](https://user-images.githubusercontent.com/1265865/206460622-e901c8ef-f9ec-4511-9941-6e355ec909cc.png)
![20221208_insert_file2](https://user-images.githubusercontent.com/1265865/206460626-09110998-53bf-4dcd-975c-d56c8ee4979c.png)

@rbovard Any suggestion how to add a check in qgep_od.data_media that the `path `contains only `/` and no `\`

**Reply by rbovard** [reply]
Better than a check, why not use the same logical as for the `vw_file` with a trigger to convert the `\` to `/` on `INSERT` and `UPDATE`?

**Reply by sjib** [reply]
What do others think? 

Could be a new file 15_data_management_functions.sql that creates a trigger like in https://github.com/QGEP/datamodel/blob/e94fef6296354464e5855be98275596af0b0c3aa/14_geometry_functions.sql#L36

**Comment by rbovard** [comment]
> I've also noticed an issue in the `vw_file`. Since `data_media.path` can be `NULL` (see [field comment](https://github.com/QGEP/datamodel/blob/master/09_qgep_dictionaries_kek.sql#L58)), the `_url` will be null in this case.

I'll open a PR to fix that.

**Reply by rbovard** [reply]
See https://github.com/QGEP/datamodel/pull/194
